### ISSUE TITLE
Version locking enhanement

### ISSUE THREAD

**Issue by Basic0** [Kopernicus/Kopernicus#299]
I understand that keeping a mod up to date is non-trivial, especially with uncertainty about upcoming KSP versions adding a risk of wasted effort.

For most mods, being out of date isn't a significant issue. Some minor functionality is lost, but the game is playable. In the case of Kopernicus, a delayed build has a significant impact on players (and being a powerful/popular mod, Kopernicus is likely to crop up in many games)

The reason stated for the lockout is completely valid - people will report bugs with no justification - but there has to be a middle-ground between being buried in bugs and losing access to long-running campaigns (potentially for months).

I'm suggesting a compromise in the form of a bypass for the lock that requires some explicit acknowledgement of the risk and acceptance of liability.

Perhaps a text box where users have to type "I accept" or similar. An alternative would be checking for the presence of a text file containing the above .

It would remove pressure from you to get a release out after a KSP build. It would allow those of us who are willing to roll back to a backup [should the need arise] to continue playing, and hopefully it will prevent spurious bug reports.

**Comment by Sigma88** [comment]
> I'm suggesting a compromise in the form of a bypass for the lock that requires some explicit acknowledgement of the risk and acceptance of liability.

this will work assuming people are reasonable when they use they bypass



people are not reasonable

**Comment by Sigma88** [comment]
also, the reason is not "people will report bugs with no justification"

first of all you are assuming that kopernicus for 1.4.3 will work fine on 1.4.4

there is not way to know that for sure, actually some people have recompiled it against 1.4.4 and said that it fails to load, so it would not help you in any way to bypass the lock

<br>
the actual reason why the lock is in place is so that kopernicus is not run on versions of ksp it is not supposed to run on. Expecially to reassure the planet makers that they can feel comfortable in releasing a planet pack on a certain version and not receive requests to update the mod for a version of kopernicus that might be flawed and that might require weird workarounds to load their planets correctly

the bypass you are proposing would basically make the lock pointless

**Comment by Basic0** [comment]
All the same... Addons not working with the latest version is a common story across the board. Kopernicus is choosing to lock people out of their save games for potentially months at a time.

If you're going to do something that drastic, you accept a measure of responsibility.

**Comment by StollD** [comment]
I prefer to lock people out of their saves for a certain amount of time, as opposed to keeping Kopernicus unlocked, and risking that they ruin their saves without a way to restore them, which is what would happen if Kopernicus was unlocked and suddenly stopped working correctly. 

This could only be solved if I would test Kopernicus immideately on every KSP update that was released and issued a public warning. But I won't be able to do that for every release, so what happens without a warning from me? People would assume that everything works just fine, load their saves, and see Kopernicus being broken and their saves ruined. This only results in tears, anger and chaos because everyone blames me (or planet pack authors) for ruining their saves.

But even with the best possible intentions, issuing a warning within hours after release just isn't possible. Checking if Kopernicus works involves updating my build infrastructure to the new KSP version, compiling Kopernicus, and then, most importantly, actually playing with it for one or two hours, and with multiple planet packs. Even better if two or three persons do that. And before something is clear, you wouldn't post anything about whether Kopernicus works or not.

And even if I that all wouldn't be important: Not all people follow the forums, not all people even know about the plumbing work Kopernicus is doing silently in the background. For those, the author of the planet pack they are using is probably their primary source of information. And the authors of the packs might be faster than me, and do a basic test if everything still works. But even overlooking the smallest issue can break the whole plugin, which is why you need to know how things work to say if they are broken.

Now assume that this planet pack author (lets call him Bill) did a quick test and saw that the planets are visible in the Tracking Station. My hypothetical self that tries to announce whether things work or not is currently on vacation without internet access, so they can't even know a new KSP version is out. Someone says thats unlikely? Just assume it for one moment. So, because Bill is sure that his planets work (he saw them after all, so they must load, right?!), he announces that Kopernicus works *himself*.

Another planet pack author (lets call him Bob) knows a bit more about how Kopernicus works - he tries to actually land on one of the bodies, and discovers that the terrain won't load at all. No matter what he tries, Kopernicus doesn't want to load it. So he rushes to his thread and warns his userbase to avoid loading their savegames.

All this is simply a result of how the KSP planet making community operates. Compared to Kopernicus staff, the planet makers simply have much more attention and publicity, because they actually make the awesome stuff. Thats different than, lets say ModuleManager, where you know that everything you made is basically because sarbian maintains ModuleManager. Planet makers are doing a gigantic (coming from someone who can only make stupid emotes in GIMP) part of their mods themselves.

So, who should the community trust now? Should they trust Bill, who says it works fine, and that they can continue their awesome career as soon as possible? Or should they trust Bob, who warns them, but they won't be able to continue their game! Now, two factors come into play: 

* as Sigma correctly stated, people are not reasonable. They might only hear what they want to hear, so they hear "Kopernicus works (with planet pack XY. pack YZ might be broken, but I am not using that, right?)"
* People might not even have contact to the same planet pack authors. Users that use Bills Mod might have no idea that Bob issued a warning.

And please understand me here, because **I can totally understand that you are fed up with me delaying the release**: This does not make any of this better. As I said previously, it creates tears, anger, and chaos on the forums. Trust me, all of this happened before. I haven't added the lock as a joke because I want to annoy someone. It is just the least harmful option that is available. Because with a version lock, the answer to "Does it work" is dead simple: **No, it doesn't**

A small note about 1.4.4: The reason for the delay is not me chilling on some island, I am actually working on improving KittopiaTech (the planet editing tool for Kopernicus) to be more stable and useable by planet makers - something I have been delaying for months now, and I finally want to get that done now. Also, I don't quite see the reason for Squad adding larger features (workshop integration) in a maintainance build, and then announcing a new maintainance release 5 days later, because they added even more bugs instead of finally fixing them. So I am more than willing to skip 1.4.4 if neccessary.

*TL;DR:* Not locking Kopernicus opens me up to a much greater amount of responsibility that I am not willing to take on me, because it might actively destroy savegames without anyone noticing until it is too late. By locking it, I am preventing users from accidentally killing their savegames without a posibility to restore them. And in exchange I am more than willing to be hated for delaying one (**!!!**) update out of many others for some weeks or months because I want to finish and polish something awesome for the planet makers, and because I am annoyed by Squads release policies. Seeing savegames break every time Kopernicus decides to break on a new KSP version (and trust me, that happened so often), in combination with the backlash that gets summoned upon the planet makers and myself would have probably made me quit KSP a long time ago.

`/rant`

**Comment by duncan-bayne** [comment]
@StollD For what it's worth, I'm impatient for the latest version, and also completely support your decision to implement a version lock (both from a technical perspective, and from a property perspective - it's your project!).

What I don't get is why Squad / Take2 would make breaking API changes between minor versions (e.g. 1.4.3 to 1.4.4).  This would be very bad form in most fields, but it seems to be an accepted norm in the game world.  From what I can see, most people _expect_ a game publisher to make breaking changes on every release.

**Comment by Sigma88** [comment]
Supposedly that option is not available for those who bought the game from the KSP store (instead of steam)

But they can still backup their games before updating, which is what they should do anyways if they use mods.

However, I would like to point out that complaining about free stuff has only one result: turning people off making more of that free stuff.

The best course of action is to be patient and **thankful** for the work people are putting into this completely for free

**Comment by marr75** [comment]
@Sigma88 truth. I have once-popular but now abandoned mods across 3 games because the comments got negative and I dreaded reading them every time. Easiest solution for me was to cease updates and cease visiting the pages.

**Comment by Basic0** [comment]
> risking that they ruin their saves without a way to restore them

In short, you believe you're better able to decide what risks people should take than they are. I fundamentally disagree with this nannying approach, but it's your code at the end of the day. I'll be sure to avoid it in future, and wish you all the best.

**Comment by Sigma88** [comment]
thank you
